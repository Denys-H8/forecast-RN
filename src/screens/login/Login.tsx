import React, { useEffect, useState } from 'react';
import { View } from 'react-native';
import * as Keychain from 'react-native-keychain';

import { styles } from './login.styles';
import Input from './components/input';
import Button from './components/button';
import constants from 'src/constants';
import { useDispatch, useSelector } from 'react-redux';
import { loginActions } from 'src/store/rootActions';
import * as LOGIN_SELECTORS from '../../store/login/selectors';

// Email: root@gmail.com
// Password: passroot

const Login: React.FC = () => {
  const [btnDisabled, setBtnDisabled] = useState(true);

  const dispatch = useDispatch();

  useEffect(() => {
    const getEmail = async () => {
      try {
        const credentials: false | Keychain.UserCredentials = await Keychain.getGenericPassword();
        if (credentials) dispatch(loginActions.changedEmail(credentials?.password));
      } catch (error) {
        console.log("Keychain couldn't be accessed!", error);
      }
    };
    getEmail();
  }, []);

  const email = useSelector(LOGIN_SELECTORS.email);
  const password = useSelector(LOGIN_SELECTORS.password);
  const emailError = useSelector(LOGIN_SELECTORS.emailError);
  const passError = useSelector(LOGIN_SELECTORS.passError);

  const emailValidation = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;

  const onChangeEmail = (val: string) => {
    dispatch(loginActions.changedEmail(val));

    emailValidation.test(val) && password.length > 7 ? setBtnDisabled(false) : setBtnDisabled(true);
  };

  const onChangePassword = (val: string) => {
    dispatch(loginActions.changedPassword(val));

    emailValidation.test(email) && val.length > 7 ? setBtnDisabled(false) : setBtnDisabled(true);
  };

  return (
    <View style={styles.container}>
      <View style={styles.content}>
        <View style={styles.topContainer}>
          <Input
            value={email}
            error={emailError}
            placeholder="Email"
            onChangeText={(val) => onChangeEmail(val)}
            placeholderTextColor={constants.MAIN_COLOR}
            iconName="account-circle"
          />
        </View>
        <View style={styles.bottomContainer}>
          <Input
            value={password}
            error={passError}
            placeholder="Password"
            onChangeText={(val) => onChangePassword(val)}
            placeholderTextColor={constants.MAIN_COLOR}
            iconName="lock"
            secureTextEntry
          />
        </View>
        <Button onPress={() => dispatch(loginActions.login())} isDisabled={btnDisabled} />
      </View>
    </View>
  );
};

export default Login;
