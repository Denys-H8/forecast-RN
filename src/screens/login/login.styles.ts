import { StyleSheet } from 'react-native';

import constants from 'src/constants';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: constants.BG_COLOR,
  },
  content: {
    flex: 1,
    paddingVertical: 50,
    paddingHorizontal: 32,
  },
  topContainer: {
    flex: 1,
    justifyContent: 'flex-end',
    paddingBottom: 12,
  },
  bottomContainer: {
    flex: 1,
    paddingTop: 12,
  },
  text: {
    color: constants.ACTIVE_TEXT_COLOR,
    fontSize: 22,
  }
});
