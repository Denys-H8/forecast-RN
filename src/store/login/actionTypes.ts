export const LOGIN = 'LOGIN';
export const LOGIN_STARTED = 'LOGIN_STARTED';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const CHANGED_EMAIL = 'CHANGED_EMAIL';
export const CHANGED_PASSWORD = 'CHANGED_PASSWORD';
export const FAILED_EMAIL = 'FAILED_EMAIL';
export const FAILED_PASSWORD = 'FAILED_PASSWORD';
export const LOGOUT = 'LOGOUT';
export const LOGOUT_SUCCESS = 'LOGOUT_SUCCESS';
export const START_LOADING = 'START_LOADING';
export const LOADING_LOGIN = 'LOADING_LOGIN';
