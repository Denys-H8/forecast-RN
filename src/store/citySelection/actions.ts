import * as ACTION from './actionTypes';
import { TGroupWeather, TData, TCity } from './reducer';

export const queryChanged = (query: string, data: TData[]) => ({
  type: ACTION.QUERY_CHANGED,
  payload: { query, data },
});
export const getWeatherForCities = () => ({ type: ACTION.GET_WEATHER_FOR_CITIES });
export const getWeatherForCitiesSucess = (data: TGroupWeather) => ({
  type: ACTION.GET_WEATHER_FOR_CITIES_SUCESS,
  payload: data,
});
export const resetQuery = () => ({ type: ACTION.RESET_QUERY });
export const changeCity = (city: TCity) => ({
  type: ACTION.CHANGE_CITY,
  payload: city,
});
export const pressedCity = (city: TCity) => ({
  type: ACTION.PRESSED_CITY,
  payload: city,
});
export const setLoading = (value: boolean) => ({ type: ACTION.SET_LOADING, payload: value });
