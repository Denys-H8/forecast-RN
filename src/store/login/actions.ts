import * as ACTION from './actionTypes';
import { TUser } from './reducer';

export const login = () => ({ type: ACTION.LOGIN });
export const logout = () => ({ type: ACTION.LOGOUT });
export const logoutSuccess = () => ({ type: ACTION.LOGOUT_SUCCESS });
export const loginStarted = () => ({ type: ACTION.LOGIN_STARTED });
export const loginSuccess = (user: TUser) => ({ type: ACTION.LOGIN_SUCCESS, payload: user });
export const changedEmail = (email: string | false) => ({ type: ACTION.CHANGED_EMAIL, payload: email });
export const changedPassword = (password: string) => ({ type: ACTION.CHANGED_PASSWORD, payload: password });
export const failedEmail = () => ({ type: ACTION.FAILED_EMAIL });
export const failedPassword = () => ({ type: ACTION.FAILED_PASSWORD });
export const startLoading = () => ({ type: ACTION.START_LOADING });
export const loading = (value: boolean) => ({ type: ACTION.LOADING_LOGIN, payload: value });
