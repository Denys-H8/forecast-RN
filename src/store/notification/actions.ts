import { FirebaseMessagingTypes } from '@react-native-firebase/messaging';
import * as ACTION from './actionTypes';

export const notification = (routeName: string) => ({ type: ACTION.NOTIFICATION, payload: routeName });
export const addNotification = (notification: FirebaseMessagingTypes.RemoteMessage) => ({
  type: ACTION.ADD_NOTIFICATION,
  payload: notification,
});
export const removeNotification = (notification: FirebaseMessagingTypes.RemoteMessage) => ({
  type: ACTION.REMOVE_NOTIFICATION,
  payload: notification,
});
