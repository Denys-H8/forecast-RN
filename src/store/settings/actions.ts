import * as ACTION from './actionTypes';

export const changedName = (name: string) => ({ type: ACTION.CHANGED_NAME, payload: name });
export const changedAge = (age: string) => ({ type: ACTION.CHANGED_AGE, payload: age });
export const changedDayPeriod = (days: number) => ({ type: ACTION.CHANGED_DAY_PERIOD, payload: days });
export const changedUpdateTime = (time: number) => ({ type: ACTION.CHANGED_UPDATE_TIME, payload: time });
export const changedTemperatureScale = (scale: string) => ({
  type: ACTION.CHANGED_TEMPERATURE_SCALE,
  payload: scale,
});
export const setTemperatureScale = (scale: string) => ({
  type: ACTION.SET_TEMPERATURE_SCALE,
  payload: scale,
});
