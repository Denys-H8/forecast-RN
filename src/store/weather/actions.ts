import * as ACTION from './actionTypes';
import { TWeather, TWeatherStats } from './reducer';

export const getWeather = () => ({ type: ACTION.GET_WEATHER });
export const getWeatherSuccess = (weatherData: TWeatherData) => ({
  type: ACTION.GET_WEATHER_SUCCESS,
  payload: weatherData,
});
export const loading = (value: boolean) => ({ type: ACTION.LOADING, payload: value });
export const lateNavigation = (routeName: string | undefined) => ({ type: ACTION.LATE_NAVIGATION, payload: routeName });
export const resetLateNavigation = () => ({ type: ACTION.RESET_LATE_NAVIGATION });

export type TWeatherData = {
  weather: TWeather;
  weatherStats: TWeatherStats;
};
