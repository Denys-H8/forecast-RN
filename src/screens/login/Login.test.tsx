import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import Login from './Login';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { changedEmail, changedPassword, login } from 'src/store/login/actions';

const mockStore = configureStore([]);

const store = mockStore({
  loginReducer: {
    email: 'email',
    password: 'password',
    emailError: false,
    passError: false,
  },
});

describe('Forecast ', () => {
  test('should render correctly', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('should handle change email', () => {
    store.dispatch = jest.fn();

    const newEmail = 'newEmail';
    const { getAllByTestId } = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );

    fireEvent.changeText(getAllByTestId('Input.TextInput')[0], newEmail);
    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(changedEmail(newEmail));
  });

  test('should handle change password', () => {
    store.dispatch = jest.fn();

    const newPassword = 'newPassword';
    const { getAllByTestId } = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );

    fireEvent.changeText(getAllByTestId('Input.TextInput')[1], newPassword);
    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(changedPassword(newPassword));
  });

  test('should handle disabled login button press', () => {
    store.dispatch = jest.fn();

    const { getByTestId } = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );

    fireEvent.press(getByTestId('Button.TouchableOpacity'));
    expect(store.dispatch).not.toHaveBeenCalled();
  });

  test('should handle not disabled login button press', () => {
    const store = mockStore({
      loginReducer: {
        email: 'email@gmail.com',
        password: 'passwordValue',
        emailError: false,
        passError: false,
      },
    });

    store.dispatch = jest.fn();

    const { getByTestId, getAllByTestId } = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );

    fireEvent.changeText(getAllByTestId('Input.TextInput')[0], 'email@gmail.com');
    fireEvent.changeText(getAllByTestId('Input.TextInput')[1], 'passwordNew');
    fireEvent(getByTestId('Button.TouchableOpacity'), 'onPress');
    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(login());
  });
});
