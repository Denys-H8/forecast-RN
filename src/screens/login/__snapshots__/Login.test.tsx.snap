// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast  should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#191C39",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 50,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-end",
          "paddingBottom": 12,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#8B8EA5",
              "borderRadius": 100,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 8,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#191C39",
                  "fontSize": 40,
                },
                undefined,
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Email"
            placeholderTextColor="#fff"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "color": "#fff",
                  "flex": 1,
                  "fontFamily": "WorkSans-Regular.ttf",
                  "fontSize": 16,
                  "lineHeight": 18,
                  "marginLeft": 8,
                },
                Object {},
              ]
            }
            testID="Input.TextInput"
            underlineColorAndroid="transparent"
            value="email"
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "paddingTop": 12,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#8B8EA5",
              "borderRadius": 100,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 8,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#191C39",
                  "fontSize": 40,
                },
                undefined,
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Password"
            placeholderTextColor="#fff"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "color": "#fff",
                  "flex": 1,
                  "fontFamily": "WorkSans-Regular.ttf",
                  "fontSize": 16,
                  "lineHeight": 18,
                  "marginLeft": 8,
                },
                Object {},
              ]
            }
            testID="Input.TextInput"
            underlineColorAndroid="transparent"
            value="password"
          />
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#7B61FF",
          "borderRadius": 100,
          "height": 48,
          "justifyContent": "center",
          "opacity": 0.4,
        }
      }
      testID="Button.TouchableOpacity"
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "WorkSans-Regular",
            },
            Object {
              "color": "#fff",
              "fontSize": 16,
              "lineHeight": 18,
            },
          ]
        }
      >
        LOGIN
      </Text>
    </View>
  </View>
</View>
`;
